readviz_prep.py

```diff
@@ -0,0 +1,142 @@
+
+
+def get_expr_for_het(mt, gq_threshold, dp_threshold):
+    return (
+        mt.GT.is_het() & (mt.GQ >= gq_threshold) & (mt.DP >= dp_threshold) & hl.is_defined(mt.meta.cram)
+    )
+
+
+def get_expr_for_hom(mt, gq_threshold, dp_threshold):
+    return (
+        mt.GT.is_diploid() & mt.GT.is_hom_var() & (mt.GQ >= gq_threshold) & (mt.DP >= dp_threshold)  & hl.is_defined(mt.meta.cram)
+    )  # & hl.is_defined(mt.meta.cram_path)
+
+
+def get_expr_for_hemi(mt, gq_threshold, dp_threshold):
+    return (
+        mt.GT.is_haploid()
+        & (mt.meta.sex == "male")
+        & (mt.GT[0] == 1)
+        & (mt.GQ >= gq_threshold)
+        & (mt.DP >= dp_threshold)
+        & hl.is_defined(mt.meta.cram)
+    )
+
+
+def main(args):
+
+    hl.init(log="/readviz_prep")
+    meta_ht = hl.read_table(args.meta_ht_path)
+    mt = hl.read_matrix_table(args.variant_mt_path).key_rows_by('locus', 'alleles')
+    crams = hl.import_table(args.cram_path).key_by('s')
+
+    dp_threshold = args.dp_threshold
+    gq_threshold = args.gq_threshold
+    num_samples = args.num_samples
+
+    meta_join = meta_ht[mt.s]
+    mt = mt.annotate_cols(
+        meta=hl.struct(
+            project_id=meta_join.project_id,
+            sex=meta_join.sex,
+            release=meta_join.release,
+            sample_filters=meta_join.sample_filters,
+            cram=crams[mt.s].final_cram_path,
+            crai=crams[mt.s].final_crai_path
+        )
+    )
+    mt = mt.filter_cols(mt.meta.release)
+    mt = hl.experimental.sparse_split_multi(mt, filter_changed_loci=True)
+
+    logger.info("Annotating genotypes with adj...")
+
+    mt = mt.annotate_entries(
+        GT=adjusted_sex_ploidy_expr(mt.locus, mt.GT, mt.meta.sex),
+        adj=get_adj_expr(mt.GT, mt.GQ, mt.DP, mt.AD)
+    )
+
+    logger.info("Selecting only entry fields needed for densify and output ...")
+    mt = mt.select_entries("GT", "GQ", "DP", "AD", "END", "adj")
+
+    if args.test:
+        mt = mt.repartition(20)
+
+    mt = hl.experimental.densify(mt)
+
+    mt = mt.filter_rows(hl.agg.any(mt.GT.is_non_ref()))
+    mt = mt.filter_rows(hl.len(mt.alleles) > 1)
+
+    logger.info(
+        f"Taking up to {num_samples} samples per variant where samples are het, hom_var, or hemi"
+    )
+    mt = mt.annotate_rows(
+        samples_w_het_var=hl.agg.filter(
+            get_expr_for_het(mt, gq_threshold, dp_threshold),
+            hl.agg.take(get_expr_for_het_hom_hemi_take(mt, "het"), num_samples, ordering=-mt.GQ),
+        ),
+        samples_w_hom_var=hl.agg.filter(
+            get_expr_for_hom(mt, gq_threshold, dp_threshold),
+            hl.agg.take(get_expr_for_het_hom_hemi_take(mt, "hom"), num_samples, ordering=-mt.GQ),
+        ),
+        samples_w_hemi_var=hl.agg.filter(
+            get_expr_for_hemi(mt, gq_threshold, dp_threshold),
+            hl.agg.take(
+                get_expr_for_het_hom_hemi_take(mt, "hemi"), num_samples, ordering=-mt.GQ
+            ),
+        ),
+    )
+    ht = mt.rows()
```

Comment:
One more request - would it be possible to add a `sample_order` field here to each struct in the `samples_w_het_var`, `samples_w_hom_var`, `samples_w_hemi_var` arrays ?
The value would  be the integer index of that struct in that specific array.
(I haven't used it, but this function might help here
https://hail.is/docs/0.2/functions/collections.html?highlight=reversed#hail.expr.functions.zip_with_index)

Reply:
Looking into adding this